dali/kernels/imgproc/convolution/laplacian_gpu.cuh

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_LAPLACIAN_GPU_CUH_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_LAPLACIAN_GPU_CUH_
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "dali/core/dev_buffer.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/block_setup.h"
+#include "dali/kernels/common/cast.cuh"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/convolution_gpu.h"
+#include "dali/kernels/imgproc/convolution/separable_convolution_gpu.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/scratch.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+
+namespace dali {
+namespace kernels {
+
+namespace laplacian {
+
+/**
+ * @brief Computes convolution to obtain partial derivative in one of the dimensions.
+ * Convolution consits of `axes` windows, each to convolve along one dimension of the input data,
+ * where `deriv_axis`-th window is supposed to compute partial derivative along that axis,
+ * whereas the remaining windows should perform smoothing. Setup method expects
+ * a `has_smoothing` hint if all window sizes in non-derivative directions are one. If so,
+ * the smoothing convolutions can be skipped and only a single one-dimensional
+ * convolution in derivative direction is performed.
+ */
+template <typename Out, typename In, typename W, int axes, int deriv_axis, bool has_channels,
+          bool is_sequence>
+struct PartialDerivGpu {
+  using MultiDimConv = SeparableConvolutionGpu<Out, In, W, axes, has_channels, is_sequence>;
+  static constexpr int ndim = MultiDimConv::ndim;
+  static constexpr int sequence_axes = MultiDimConv::sequence_axes;
+  using SingleDimConv = ConvolutionGpu<Out, In, W, ndim, sequence_axes + deriv_axis, has_channels>;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const std::array<TensorListShape<1>, axes>& window_sizes,
+                           bool has_smoothing) {
```

Comment:
<discussed, offline; here for visibility>: I think that has_smoothing could be derived from `window_size` being empty. This would allow specifying this behavior per-axis from LaplacianGpu level.

Reply:
done